node/service/src/parachains_db/upgrade.rs

```diff
@@ -467,6 +546,77 @@ mod tests {
 		);
 	}
 
+	#[test]
+	fn test_migrate_3_to_4() {
+		use kvdb_rocksdb::{Database, DatabaseConfig};
+		use polkadot_node_core_approval_voting::approval_db::v2::migration_helpers::v1_to_v2_sanity_check;
+		use polkadot_node_subsystem_util::database::kvdb_impl::DbAdapter;
+
+		let db_dir = tempfile::tempdir().unwrap();
+		let db_path = db_dir.path().to_str().unwrap();
+		let db_cfg: DatabaseConfig = DatabaseConfig::with_columns(super::columns::v3::NUM_COLUMNS);
+
+		let approval_cfg = ApprovalDbConfig {
+			col_approval_data: crate::parachains_db::REAL_COLUMNS.col_approval_data,
+		};
+
+		// We need to properly set db version for upgrade to work.
+		fs::write(version_file_path(db_dir.path()), "3").expect("Failed to write DB version");
+		let expected_candidates = {
+			let db = Database::open(&db_cfg, db_path.clone()).unwrap();
+			assert_eq!(db.num_columns(), super::columns::v3::NUM_COLUMNS as u32);
+			let db = DbAdapter::new(db, columns::v3::ORDERED_COL);
+			// Fill the approval voting column with test data.
+			v1_to_v2_fill_test_data(std::sync::Arc::new(db), approval_cfg.clone()).unwrap()
+		};
+
+		try_upgrade_db(&db_dir.path(), DatabaseKind::RocksDB, 4).unwrap();
+
+		let db_cfg = DatabaseConfig::with_columns(super::columns::v4::NUM_COLUMNS);
+		let db = Database::open(&db_cfg, db_path).unwrap();
+		let db = DbAdapter::new(db, columns::v4::ORDERED_COL);
+
+		v1_to_v2_sanity_check(std::sync::Arc::new(db), approval_cfg.clone(), expected_candidates)
+			.unwrap();
+	}
+
+	#[test]
+	fn test_rocksdb_migrate_0_to_4() {
+		use kvdb_rocksdb::{Database, DatabaseConfig};
+
+		let db_dir = tempfile::tempdir().unwrap();
+		let db_path = db_dir.path().to_str().unwrap();
+
+		fs::write(version_file_path(db_dir.path()), "0").expect("Failed to write DB version");
+		try_upgrade_db(&db_dir.path(), DatabaseKind::RocksDB, 4).unwrap();
+
+		let db_cfg = DatabaseConfig::with_columns(super::columns::v4::NUM_COLUMNS);
+		let db = Database::open(&db_cfg, db_path).unwrap();
+
+		assert_eq!(db.num_columns(), columns::v4::NUM_COLUMNS);
+	}
+
+	#[test]
+	fn test_paritydb_migrate_0_to_4() {
```

Comment:
this needs to be checked on flakyness after https://github.com/paritytech/parity-db/pull/218